src/librustc_metadata/decoder.rs

```diff
@@ -917,22 +917,32 @@ impl<'a, 'tcx> CrateMetadata {
         self.maybe_entry(id).and_then(|item| item.decode(self).mir).is_some()
     }
 
-    pub fn maybe_get_optimized_mir(&self, tcx: TyCtxt<'tcx>, id: DefIndex) -> Option<Body<'tcx>> {
-        match self.is_proc_macro(id) {
-            true => None,
-            false => self.entry(id).mir.map(|mir| mir.decode((self, tcx))),
-        }
+    pub fn get_optimized_mir(&self, tcx: TyCtxt<'tcx>, id: DefIndex) -> Body<'tcx> {
+        let mir =
```

Comment:
This overall pattern:
```rust
        match self.is_proc_macro(id) {
            true => fallback(),
            false => stuff(self.entry(id)),
        }
```
occurs a few times in this file (search for `self.is_proc_macro(`).

Consider adding a helper:
```rust
fn entry_unless_proc_macro(&self, id: _) -> Option<_> {
    match self {
        true => None,
        _ => self.entry(id),
    }
}
```
and use it as:
```rust
    self.entry_unless_proc_macro(id)
        .unwrap_or_else(|| { ... })
        .mir
        .decode((self, tcx))
```